Hybrid Nodes Networking Docs (pod routing, webhooks, mixed mode clusters)

csplinter opened the pull request:
*Description of changes:*
This doc update adds to the info we have for EKS Hybrid Nodes on pod network routing, webhooks, and running mixed mode clusters

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

csplinter committed:
add content for hybrid nodes pod network routability and mixed mode clusters

csplinter committed:
CNI docs improvements, add considerations section

csplinter committed:
use correct formats for code snippets

csplinter committed:
more formatting changes for replaceable fields

g-gaston commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -63,23 +63,32 @@ Cilium version `1.16.x` is supported and recommended for EKS Hybrid Nodes for ev
-A common way to advertise pod addresses with your on-premises network is by using BGP. To use BGP with Cilium, you must set `bgpControlPlane.enabled: true`. For more information on Cilium's BGP support, see https://docs.cilium.io/en/stable/network/bgp-control-plane/bgp-control-plane/[Cilium BGP Control Plane] in the Cilium documentation.
+. Create a YAML file called `cilium-values.yaml`. The following example configures Cilium to run on hybrid nodes only by setting affinity for the `eks.amazonaws.com/compute-type: hybrid` label. 
+
+- If you configured your Amazon EKS cluster with _remote pod networks_, configure the same pod CIDRs for your  `clusterPoolIPv4PodCIDRList`. For example, `10.100.0.0/24`.
```
Worth asking to validate that POD_CIDR doesn't overlap with the node cidr?

Even if pod cidr is not configured at the eks level, the pod cidr cilium uses (regardless if manually input or the default one) must not overlap with the node cidr.

csplinter replied:
ah yea good callout. will add that

g-gaston commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -338,6 +350,14 @@ The interfaces and routes configured by Cilium are not removed by default when t
+- It is recommended to run Calico in overlay / tunnel mode with VXLAN as the link:https://docs.tigera.io/calico/latest/networking/configuring/vxlan-ipip[encapsulation method]. This mode has the fewest requirements on the underlying physical network. For more information on the different Calico networking modes, see https://docs.tigera.io/calico/latest/networking/determine-best-networking[Determin ...
+- It is recommended to run Calico with `natOutgoing` set to `true`. With `natOutgoing` set to `true` the source IP address of all pod traffic leaving the cluster to the IP address of the node. This makes it possible to run Calico with Amazon EKS clusters that have remote pod networks configured and with clusters that don't have remote pod networks configured. If you disable `natOutgoing`, then yo ...
+- If you are running webhooks on your hybrid nodes, your pod CIDRs must be routable on your on-premises network and you must configure your Amazon EKS cluster with your remote pod networks. If your pod CIDRs are not routable on your on-premises network, then it is recommended to run webhooks on cloud nodes in the same cluster. See <<hybrid-nodes-networking#mixed-mode-clusters, Mixed mode clusters ...
+- - A common way to make your pod CIDR routable on your on-premises network is to advertise pod addresses with BGP. To use BGP with Calico, you must set `installation.calicoNetwork.bgp: Enabled` in your Helm configuration. For more information on Calico's BGP support, see link:https://docs.tigera.io/calico/latest/networking/configuring/bgp[Configure BGP peering] in the Calico documentation.
```
```suggestion
- A common way to make your pod CIDR routable on your on-premises network is to advertise pod addresses with BGP. To use BGP with Calico, you must set `installation.calicoNetwork.bgp: Enabled` in your Helm configuration. For more information on Calico's BGP support, see link:https://docs.tigera.io/calico/latest/networking/configuring/bgp[Configure BGP peering] in the Calico documentation.
```

erchpm commented on latest/ug/nodes/hybrid-nodes-add-ons.adoc:
```diff
@@ -62,7 +62,67 @@ The sections that follow describe differences between running compatible {aws} a
+
+If you are running a mixed mode cluster with both hybrid nodes and nodes in {aws} Cloud, we recommend that you have at least one CoreDNS replica on hybrid nodes and at least one CoreDNS replica on your nodes in {aws} Cloud. CoreDNS can be configured such that your workloads will use the closest CoreDNS replica, meaning your cloud workloads will use the CoreDNS running in the cloud and your hybrid ...
+
+. Add a topology zone label for each of your hybrid nodes, for example `topology.kubernetes.io/zone: onprem`. This can alternatively be done at the `nodeadm init` phase. Note, nodes running in {aws} Cloud automatically get a topology zone label applied to them.
```
Is the topology label added at the `nodeadm init` by including a label to the NodeConfig spec? If yes, it could help to call that out.

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -164,11 +176,11 @@ spec:
-      peerASN: PEER_ASN
-      peerAddress: ONPREM_ROUTER_IP
+      peerASN: [.replaceable]`PEER_ASN`
+      peerAddress: [.replaceable]`ONPREM_ROUTER_IP`
```
Can we provide guidance on how to find localASN, peerASN, and peerAddress, or at least what those values represent? I've always been the most ucerain at this step.

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -338,6 +350,14 @@ The interfaces and routes configured by Cilium are not removed by default when t
+== Calico considerations
+
+- It is recommended to run Calico in overlay / tunnel mode with VXLAN as the link:https://docs.tigera.io/calico/latest/networking/configuring/vxlan-ipip[encapsulation method]. This mode has the fewest requirements on the underlying physical network. For more information on the different Calico networking modes, see https://docs.tigera.io/calico/latest/networking/determine-best-networking[Determin ...
+- It is recommended to run Calico with `natOutgoing` set to `true`. With `natOutgoing` set to `true` the source IP address of all pod traffic leaving the cluster to the IP address of the node. This makes it possible to run Calico with Amazon EKS clusters that have remote pod networks configured and with clusters that don't have remote pod networks configured. If you disable `natOutgoing`, then yo ...
```
```suggestion
- It is recommended to run Calico with `natOutgoing` set to `true`. With `natOutgoing` set to `true` the source IP address of all pod traffic leaving the cluster is translated to the IP address of the node. This makes it possible to run Calico with Amazon EKS Hybrid Nodes, whether or not remote pod networks are configured on the cluster. If you disable `natOutgoing`, then your pod CIDRs must be routable on your on-premises network and you must configure your Amazon EKS cluster with your remote pod networks. 
```

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -338,6 +350,14 @@ The interfaces and routes configured by Cilium are not removed by default when t
+- It is recommended to run Calico with `natOutgoing` set to `true`. With `natOutgoing` set to `true` the source IP address of all pod traffic leaving the cluster to the IP address of the node. This makes it possible to run Calico with Amazon EKS clusters that have remote pod networks configured and with clusters that don't have remote pod networks configured. If you disable `natOutgoing`, then yo ...
+- If you are running webhooks on your hybrid nodes, your pod CIDRs must be routable on your on-premises network and you must configure your Amazon EKS cluster with your remote pod networks. If your pod CIDRs are not routable on your on-premises network, then it is recommended to run webhooks on cloud nodes in the same cluster. See <<hybrid-nodes-networking#mixed-mode-clusters, Mixed mode clusters ...
+- - A common way to make your pod CIDR routable on your on-premises network is to advertise pod addresses with BGP. To use BGP with Calico, you must set `installation.calicoNetwork.bgp: Enabled` in your Helm configuration. For more information on Calico's BGP support, see link:https://docs.tigera.io/calico/latest/networking/configuring/bgp[Configure BGP peering] in the Calico documentation.
+- The default IP Address Management (IPAM) in Calico is called link:https://docs.tigera.io/calico/latest/networking/ipam/get-started-ip-addresses#calico-ipam[Calico IPAM], where the `calico-ipam` plugin allocates IP addresses for each node based on user-configured pod CIDRs. The pod CIDRs are  configured with the `installation.calicoNetwork.ipPools.cidr` Helm value. Calico allocates segments from ...
```
```suggestion
- The default IP Address Management (IPAM) in Calico is called link:https://docs.tigera.io/calico/latest/networking/ipam/get-started-ip-addresses#calico-ipam[Calico IPAM], where the `calico-ipam` plugin allocates IP addresses for each node based on user-configured pod CIDRs. The pod CIDRs are  configured with the `installation.calicoNetwork.ipPools.cidr` Helm value, which should match the remote pod network CIDRs you configured for your Amazon EKS cluster. Calico allocates segments from the `ipPools.cidr` to each node. The size of the per node segments is configured with the `ipPools.blockSize` Helm value. For more information on IPAM with Calico, see link:https://docs.tigera.io/calico/latest/networking/ipam/get-started-ip-addresses[Get started with IP address management] in the Calico documentation.
```

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -348,12 +368,12 @@ kubectl get crds -oname | grep "cilium" | xargs kubectl delete
-.. Replace `POD_CIDR` with the CIDR ranges for your pods. If you configured your Amazon EKS cluster with remote pod networks, the `POD_CIDR` that you specify for Calico should be the same as the remote pod networks. For example, `10.100.0.0/24`.
-.. Replace `CIDR_SIZE` with the size of the CIDR segment you want to allocate to each node. For example, `25` for a /25 segment size. For more information on CIDR `blockSize` and changing the `blockSize`, see https://docs.tigera.io/calico/latest/networking/ipam/change-block-size[Change IP pool block size] in the Calico documentation.
-.. In the example below, `natOutgoing` is enabled and `bgp` is disabled. In this configuration, Calico can run on Amazon EKS clusters that have Remote Pod Network configured and can run on clusters that do not have Remote Pod Network configured. If you have `natOutgoing` set to disabled, you must configure your cluster with your remote pod networks and your on-premises network must be able to pro ...
+. Create a YAML file called `calico-values.yaml`. The following example configures all Calico components to run on hybrid nodes only by setting affinity for the `eks.amazonaws.com/compute-type: hybrid` label. 
```
```suggestion
. Create a YAML file called `calico-values.yaml`. The following example configures all Calico components to run only on hybrid nodes by setting affinity for the `eks.amazonaws.com/compute-type: hybrid` label. 
```

erchpm commented on latest/ug/nodes/hybrid-nodes-networking.adoc:
```diff
@@ -25,14 +25,24 @@ For an optimal experience, {aws} recommends reliable network connectivity of at
+
+Mixed mode clusters are defined as EKS clusters that have both hybrid and cloud nodes. When running a mixed mode cluster, it is recommended to run the VPC CNI on cloud nodes and either Cilium or Calico on hybrid nodes. Cilium and Calico are not supported by AWS when running on cloud nodes. It is recommended to run at least one replica of CoreDNS on cloud nodes and at least one replica of CoreDNS  ...
+
+If your applications require pods running on cloud nodes to directly communicate with pods running on hybrid nodes ("east-west communication"), then your pod CIDR must be routable on your on-premises network when using the VPC CNI on cloud nodes and Cilium or Calico in overlay/tunnel mode on hybrid nodes. In this setup, the pods in the cloud get IP addresses from the VPC CNI based on the VPC subn ...
```
```suggestion
If your applications require pods running on cloud nodes to directly communicate with pods running on hybrid nodes ("east-west communication") and you are using the VPC CNI on cloud nodes and Cilium or Calico in overlay/tunnel mode on hybrid nodes, then your pod CIDR must be routable on your on-premises network. In this setup, the pods in the cloud get IP addresses from the VPC CNI based on the VPC subnet configuration and communicate over the VPC network. If your on-premises pod CIDR range is configured in your VPC routing table with a route to your gateway (TGW or VGW), then the traffic destined for pods on hybrid nodes will be routed to the gateway and on to your on-premises network. Your pod CIDR must be routable on your on-premises network, because when the traffic reaches your on-premises network, your router must know which hybrid node owns the IP address space of the destination pod.
```

erchpm commented on latest/ug/nodes/hybrid-nodes-add-ons.adoc:
```diff
@@ -62,7 +62,67 @@ The sections that follow describe differences between running compatible {aws} a
 == kube-proxy and CoreDNS
 
-EKS installs Kube-proxy and CoreDNS as self-managed add-ons by default when you create an EKS cluster with the {aws} API and {aws} SDKs, including from the {aws} CLI. You can overwrite these add-ons as Amazon EKS add-ons after cluster creation. Reference the EKS documentation for details on <<managing-kube-proxy>> and <<managing-coredns>>. If you are running a cluster with hybrid nodes and nodes  ...
+EKS installs kube-proxy and CoreDNS as self-managed add-ons by default when you create an EKS cluster with the {aws} API and {aws} SDKs, including from the {aws} CLI. You can overwrite these add-ons as Amazon EKS add-ons after cluster creation. Reference the EKS documentation for details on <<managing-kube-proxy>> and <<managing-coredns>>. 
```
```suggestion
EKS installs kube-proxy and CoreDNS as self-managed add-ons by default when you create an EKS cluster with the {aws} API and {aws} SDKs, including from the {aws} CLI. You can overwrite these add-ons with Amazon EKS add-ons after cluster creation. Reference the EKS documentation for details on <<managing-kube-proxy>> and <<managing-coredns>>. 
```

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -63,23 +63,32 @@ Cilium version `1.16.x` is supported and recommended for EKS Hybrid Nodes for ev
 
+== Cilium considerations
+
+- By default, Cilium is configured to run in overlay / tunnel mode with VXLAN as the link:https://docs.cilium.io/en/stable/network/concepts/routing/#encapsulation[encapsulation method]. This mode has the fewest requirements on the underlying physical network.
```
```suggestion
- By default, Cilium is configured to run in overlay / tunnel mode with VXLAN as the link:https://docs.cilium.io/en/stable/network/concepts/routing/#encapsulation[encapsulation method]. This mode is the most flexible to the configuration of the underlying physical network..
```

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -63,23 +63,32 @@ Cilium version `1.16.x` is supported and recommended for EKS Hybrid Nodes for ev
+== Cilium considerations
+
+- By default, Cilium is configured to run in overlay / tunnel mode with VXLAN as the link:https://docs.cilium.io/en/stable/network/concepts/routing/#encapsulation[encapsulation method]. This mode has the fewest requirements on the underlying physical network.
+- By default, Cilium link:https://docs.cilium.io/en/stable/network/concepts/masquerading/[masquerades] the source IP address of all pod traffic leaving the cluster to the IP address of the node. This makes it possible to run Cilium with Amazon EKS clusters that have remote pod networks configured and with clusters that don't have remote pod networks configured. If you disable masquerading, then y ...
```
```suggestion
- By default, Cilium link:https://docs.cilium.io/en/stable/network/concepts/masquerading/[masquerades] the source IP address of all pod traffic leaving the cluster to the IP address of the node. This makes it possible to run Cilium with Amazon EKS Hybrid Nodes whether or not remote pod networks are configured on the cluster. If you disable masquerading, then your pod CIDRs must be routable on your on-premises network and you must configure your Amazon EKS cluster with your remote pod networks. 
```

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -63,23 +63,32 @@ Cilium version `1.16.x` is supported and recommended for EKS Hybrid Nodes for ev
+- By default, Cilium link:https://docs.cilium.io/en/stable/network/concepts/masquerading/[masquerades] the source IP address of all pod traffic leaving the cluster to the IP address of the node. This makes it possible to run Cilium with Amazon EKS clusters that have remote pod networks configured and with clusters that don't have remote pod networks configured. If you disable masquerading, then y ...
+- If you are running webhooks on your hybrid nodes, your pod CIDRs must be routable on your on-premises network and you must configure your Amazon EKS cluster with your remote pod networks. If your pod CIDRs are not routable on your on-premises network, then it is recommended to run webhooks on cloud nodes in the same cluster. See <<hybrid-nodes-networking#mixed-mode-clusters, Mixed mode clusters ...
+- A common way to make your pod CIDR routable on your on-premises network is to advertise pod addresses with BGP. To use BGP with Cilium, you must set `bgpControlPlane.enabled: true` in your Helm configuration. For more information on Cilium's BGP support, see https://docs.cilium.io/en/stable/network/bgp-control-plane/bgp-control-plane/[Cilium BGP Control Plane] in the Cilium documentation.
+- The default IP Address Management (IPAM) in Cilium is called link:https://docs.cilium.io/en/stable/network/concepts/ipam/cluster-pool/[Cluster Scope], where the Cilium operator allocates IP addresses for each node based on user-configured pod CIDRs. The pod CIDRs are  configured with the `clusterPoolIPv4PodCIDRList` Helm value. Cilium allocates segments from the `clusterPoolIPv4PodCIDRList` to  ...
```
```suggestion
- The default IP Address Management (IPAM) in Cilium is called link:https://docs.cilium.io/en/stable/network/concepts/ipam/cluster-pool/[Cluster Scope], where the Cilium operator allocates IP addresses for each node based on user-configured pod CIDRs. The pod CIDRs are configured with the `clusterPoolIPv4PodCIDRList` Helm value, which should match the remote pod network CIDRs you configured for your Amazon EKS cluster. Cilium allocates segments from the `clusterPoolIPv4PodCIDRList` to each node. The size of the per node segments is configured with the `clusterPoolIPv4MaskSize` Helm value. For more information on the `clusterPoolIPv4PodCIDRList` and `clusterPoolIPv4MaskSize`, see https://docs.cilium.io/en/stable/network/concepts/ipam/cluster-pool/#expanding-the-cluster-pool[Expanding the cluster pool] in the Cilium documentation.
```

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -63,23 +63,32 @@ Cilium version `1.16.x` is supported and recommended for EKS Hybrid Nodes for ev
-By default, Cilium masquerades the source IP address of all pod traffic leaving the cluster to the IP address of the node. This makes it possible for Cilium to run with Amazon EKS clusters that have remote pod networks configured and with clusters that don't have remote pod networks configured. If you disable masquerading for your Cilium deployment, then you must configure your Amazon EKS cluster ...
-+
-A common way to advertise pod addresses with your on-premises network is by using BGP. To use BGP with Cilium, you must set `bgpControlPlane.enabled: true`. For more information on Cilium's BGP support, see https://docs.cilium.io/en/stable/network/bgp-control-plane/bgp-control-plane/[Cilium BGP Control Plane] in the Cilium documentation.
+. Create a YAML file called `cilium-values.yaml`. The following example configures Cilium to run on hybrid nodes only by setting affinity for the `eks.amazonaws.com/compute-type: hybrid` label. 
```
```suggestion
. Create a YAML file called `cilium-values.yaml`. The following example configures Cilium to run only on hybrid nodes by setting affinity for the `eks.amazonaws.com/compute-type: hybrid` label. 
```

erchpm commented on latest/ug/nodes/hybrid-nodes-cni.adoc:
```diff
@@ -338,6 +350,14 @@ The interfaces and routes configured by Cilium are not removed by default when t
 
+== Calico considerations
+
+- It is recommended to run Calico in overlay / tunnel mode with VXLAN as the link:https://docs.tigera.io/calico/latest/networking/configuring/vxlan-ipip[encapsulation method]. This mode has the fewest requirements on the underlying physical network. For more information on the different Calico networking modes, see https://docs.tigera.io/calico/latest/networking/determine-best-networking[Determin ...
```
```suggestion
- It is recommended to run Calico in overlay / tunnel mode with VXLAN as the link:https://docs.tigera.io/calico/latest/networking/configuring/vxlan-ipip[encapsulation method]. This mode is the most flexible to the configuration of the underlying physical network. For more information on the different Calico networking modes, see https://docs.tigera.io/calico/latest/networking/determine-best-networking[Determining the best networking option] in the Calico documentation.
```

csplinter committed:
create separate page for hybrid nodes and webhooks

csplinter committed:
minor additions, link to mixed-mode sections from overview

csplinter commented:
This is ready for docs team review now

csplinter committed:
add commands to see webhooks on cluster

fincd-aws committed:
Fix CI breaks in PR 906

 Don't write out 'AWS' literally. Use the variable '{aws}' instead.

 Added explicit IDs to the add-on sections in hybrid-nodes-webhooks to
avoid collisions between auto-generated IDs and the IDs of the main
sections for each add-on.

fincd-aws commented:
ok, that rebase wasn't good. I'll see what I can do...

csplinter committed:
minor cleanup to make wording and links consistent

csplinter committed:
add links to webhooks docs

fincd-aws committed:
Fix links and branding in PR 906
